pkg/codegen/go/gen_program.go

```diff
@@ -54,22 +69,38 @@ func GenerateProgram(program *hcl2.Program) (map[string][]byte, hcl.Diagnostics,
 		optionalSpiller:     &optionalSpiller{},
 		scopeTraversalRoots: codegen.NewStringSet(),
 		arrayHelpers:        make(map[string]*promptToInputArrayHelper),
+		importAliases:       importAliases,
+		importBasePaths:     importBasePaths,
+		modToPkg:            modToPkg,
 	}
 
 	g.Formatter = format.NewFormatter(g)
 
-	var index bytes.Buffer
-	g.genPreamble(&index, program)
+	// we must collect imports once before lowering, and once after.
+	// this allows us to avoid complexity of traversing apply expressions for things like JSON
+	// but still have access to types provided by __convert intrinsics after lowering.
+	pulumiImports := codegen.NewStringSet()
+	stdImports := codegen.NewStringSet()
+	g.collectImports(program, stdImports, pulumiImports)
 
+	var progPostamble bytes.Buffer
 	for _, n := range nodes {
 		g.collectScopeRoots(n)
 	}
 
 	for _, n := range nodes {
-		g.genNode(&index, n)
+		g.genNode(&progPostamble, n)
 	}
 
-	g.genPostamble(&index, nodes)
+	g.genPostamble(&progPostamble, nodes)
+
+	// We must genearte the program first and the preamble second and finally cat the two together.
+	// This is because nested object/tuple cons expressions can require imports that aren't
+	// present in resource declarations or invokes alone. Expressions are lowered when the program is generated
+	// and this must happen first so we can access types via __convert intrinsics.
```

Comment:
Note that we could also approach this by lowering the entire program prior to generation. I think the approach you've taken here is fine, though.

Reply:
Yes, but unfortunately lowering the program has side-effects in the form of producing temp variables that must be emitted inline as code is generated. 

This scenario makes it obvious that this is not a great design, but nothing much to do about it at this point in time.